### PR TITLE
Add empty test files for packages with no test files

### PR DESCRIPTION
**Pull request by peter-edge**
https://github.com/golang/go/issues/6725

If you look at codecov, and look at the numbers, if a package has no tests and is not included in any package that has a test, it doesn't count towards code coverage metrics. The goal is to measure code coverage metrics only for production library code, but also for ALL production library code - once we have this, codecov will be as useful as we can make it.

These are the packages we do not ignore, but do not have a test file in them, as of right now:

```
api/middleware/middlewaretest
api/peer
api/peer/peertest
api/transport/transporttest
encoding/x/protobuf/protoc-gen-yarpc-go
internal/encoding
internal/introspection
internal/pally/pallytest
internal/protoplugin
internal/request
internal/routertest
internal/whitespace
transport/x/cherami/internal
transport/x/cherami/mocks
transport/x/redis/redistest
```

Note the *test packages (`middlewaretest`, `transporttest`, etc) arguably should be ignored as well, and I believe codecov does this automatically (they don't seem to appear in codecov, even after adding a marker file, and do not seem to affect code coverage metrics). If we want to add these to the ignore list, we should separately, however technically they contain code we give to customers.

### CODE REVIEW

**Comment by peter-edge** [comment]
Done

**Comment by breerly** [comment]
Since Go 1.2... closed due to inactivity :/

**Comment by peter-edge** [comment]
Note this will be an issue with all golang libs, cc'ing @sectioneight if you have any interest

**Comment by peter-edge** [comment]
Also @prashantv

**Comment by peter-edge** [comment]
Note the *test packages (`middlewaretest`, `transporttest`, etc) arguably should be ignored as well, and I believe codecov does this automatically (they don't seem to appear in codecov, even after adding a marker file). If we want to add these to the ignore list, we should separately, however technically they contain code we give to customers.

**Comment by peter-edge** [comment]
@breerly what is your opinion on adding all packages that end in *test to ignores? Most of them are mocks, they are arguably not production lib code.

**Comment by sectioneight** [comment]
Have you guys tried `go test -coverpkg`? I think that's supposed to catch these things and also note coverage if you test it from another package but not the package the unit test is in.

**Comment by peter-edge** [comment]
@sectioneight we have that, but that will only catch files if they are tested in a test in another package, ie if `a/b/c.go` is included in `a/d.go`, which has `a/d_test.go` that calls code in `a/b/c.go`.

**Comment by peter-edge** [comment]
Aaaand you pointed that out, yea :) The point being, there are packages that aren't even tested in OTHER packages, and they are blatantly missed.

**Comment by peter-edge** [comment]
@breerly after running code coverage, it turns out that the only missing packages are in fact `internal/protoplugin` and `encoding/x/protobuf/protoc-gen-yarpc-go` - everything else ends up being called by test code outside of their packages. Do we want to just add a test marker for these packages, or add one for all? I think we should still have a marker for all, we can't rely on this, but I don't know.

**Comment by akshayjshah** [comment]
Let's prioritize discussing this immediately after standup today.

**Comment by willhug** [comment]
If this is the direction going forward, can we add a linter to verify every go package has a test file?

**Comment by peter-edge** [comment]
Ya, I'll do that as part of this PR tomorrow, good call

File: scripts/check-test-files.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    if [ -z "${NO_TEST_FILE_DIRS}" ]; then
+      NO_TEST_FILE_DIRS="${dir}"
+    else
+      NO_TEST_FILE_DIRS="${NO_TEST_FILE_DIRS} ${DIR}"
+    fi
+  fi
+done
+
+if [ -n "${NO_TEST_FILE_DIRS}" ]; then
+  echo "error: at least one _test.go file must be in these directories: ${NO_TEST_FILE_DIRS}" >&2
```
Comment (willhug) [inline scripts/check-test-files.sh]: we should specify this is for coverage.

Comment (peter-edge) [inline scripts/check-test-files.sh]: Creating an empty test file is more difficult - you have to detect the package name. I can do this, but do you think this is necessary? I rather do this in a separate diff.

File: scripts/check-test-files.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+set -euo pipefail
+
+DIR="$(cd "$(dirname "${0}")/.." && pwd)"
+source "${DIR}/scripts/source.sh"
```
Comment (willhug) [inline scripts/check-test-files.sh]: \<nit\> source doesn't really provide meaning, maybe `directory-helpers`?

**Pull request merged**